# This file runs doubletfinder for single-cell rna-seq data 

# Doubletfinder cannot be run on merged samples! Samples must be run through doubletfinder first and then merged.
# Have to do basic QC before using doubletfinder :

```{r}
remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
```
# Load libraries
```{r}
library(DoubletFinder)
library(Seurat)
library(tidyverse)
library(ggplot2)
```

# Load in data
```{r}
CH_DEV <- Read10X_h5("/Users/aakritisingh/Documents/MZM_genome_10x_out_files/CH_DEV_filtered_feature_bc_matrix.h5")
so_CH_DEV <- CreateSeuratObject(counts = CH_DEV, project = "CH_Development")
so_CH_DEV
CH_DIA <- Read10X_h5("/Users/aakritisingh/Documents/MZM_genome_10x_out_files/CH_DIA_filtered_feature_bc_matrix.h5")
so_CH_DIA <- CreateSeuratObject(counts = CH_DIA, project = "CH_Diapause")
so_CH_DIA
SG_DEV <- Read10X_h5("/Users/aakritisingh/Documents/MZM_genome_10x_out_files/SG_DEV_filtered_feature_bc_matrix.h5")
so_SG_DEV <- CreateSeuratObject(counts = SG_DEV, project = "SG_Development")
so_SG_DEV
SG_DIA <- Read10X_h5("/Users/aakritisingh/Documents/MZM_genome_10x_out_files/SG_DIA_filtered_feature_bc_matrix.h5")
so_SG_DIA <- CreateSeuratObject(counts = SG_DIA, project = "SG_Diapause")
so_SG_DIA
```


# ------------------------------------------------------------------------------------------------------------------------------------------
# SEARCHING FOR MITOCHONDRIAL GENES FOR PREPROCESSING
# - In ncbi GTF, they are written with underscore _
# - In seurat object, they are changed to have a dash -
# - List of mitochondrial genes: nd1, nd2, cox1, cox2, atp8, atp6, cox3, nd3, nd4L, nd4, nd5, nd6, cytB

```{r}
mito.features <- c("KEG92-p13", "KEG92-p12", "KEG92-p11", "KEG92-p10", "KEG92-p09", "KEG92-p08", "KEG92-p07", "KEG92-p06", "KEG92-p05", "KEG92-p04", "KEG92-p03", "KEG92-p02", "KEG92-p01")
all(mito.features %in% rownames(so_CH_DEV))
```

```{r}
so_CH_DEV[["RNA"]]
```
```{r}
View(so_CH_DEV@assays$RNA$`counts.Gene Expression`)
so_CH_DEV@assays$RNA$`counts.Gene Expression`
```


```{r}
so_CH_DEV
test <- WhichCells(so_CH_DEV, expression = 'Geneid' > 0)
length(test)
```

# -----------------------------------------------------------------------------------------------------------------------------------------
# SEURAT QC AND PREPROCESSING

# Provide list of mitochondrial genes for preprocessing.
# From GTF file, genes are named as:
# KEG92_p13 [ND1] KEG92-p13
# KEG92_p12 [ND2] KEG92-p12
# KEG92_p11 [COX1] KEG92-p11
# KEG92_p10 [COX2] KEG92-p10
# KEG92_p09 [ATP8] KEG92-p09
# KEG92_p08 [ATP6] KEG92-p08
# KEG92_p07 [COX3] KEG92-p07
# KEG92_p06 [ND3] KEG92-p06
# KEG92_p05 [ND4L] KEG92-p05
# KEG92_p04 [ND4] KEG92-p04
# KEG92_p03 [ND5] KEG92-p03
# KEG92_p02 [ND6] KEG92-p02
# KEG92_p01 [CYTB] KEG92-p01


```{r}
so_CH_DEV[["RNA"]]$counts.Peaks <- NULL
```

```{r}
so_CH_DEV[["percent.mt"]] <- PercentageFeatureSet(so_CH_DEV, features = c("KEG92-p13", "KEG92-p12", "KEG92-p11", "KEG92-p10", "KEG92-p09", "KEG92-p08", "KEG92-p07", "KEG92-p06", "KEG92-p05", "KEG92-p04", "KEG92-p03", "KEG92-p02", "KEG92-p01"), assay = 'RNA')
VlnPlot(so_CH_DEV, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", ncol=3))
```

```{r}
CH_DEV <- Read10X_h5("/Users/aakritisingh/Documents/MZM_genome_10x_out_files/CH_DEV_filtered_feature_bc_matrix.h5")
so_CH_DEV <- CreateSeuratObject(counts = CH_DEV, project = "CH_Development")
so_CH_DEV
VlnPlot(so_CH_DEV, features = c("nFeature_RNA"))
```
```{r}
so_CH_DEV[["RNA"]]$counts.Peaks <- NULL
VlnPlot(so_CH_DEV, features = c("nFeature_RNA"))
matrix <- so_CH_DEV$nFeature_RNA 
matrix <- as.data.frame(matrix)
write_csv(matrix, "/Users/aakritisingh/Downloads/matrix2.csv")
```

Seurat filtering

```{r}
so_CH_DEV_filtered <- subset(so_CH_DEV, subset = nFeature_RNA > 200 & nFeature_RNA < 1000 & percent.mt < 10)
VlnPlot(so_CH_DEV_filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
```

-----------------------------------------------------------------------------------------------------------------------------------------
# After QC and subsetting, perform basic seurat workflow

```{r}
so_CH_DEV_filtered <- NormalizeData(object = so_CH_DEV_filtered)
so_CH_DEV_filtered <- FindVariableFeatures(object = so_CH_DEV_filtered)
so_CH_DEV_filtered <- ScaleData(object = so_CH_DEV_filtered)
so_CH_DEV_filtered <- RunPCA(object = so_CH_DEV_filtered)
ElbowPlot(so_CH_DEV_filtered)
so_CH_DEV_filtered <- FindNeighbors(object = so_CH_DEV_filtered, dims = 1:20)
so_CH_DEV_filtered <- FindClusters(object = so_CH_DEV_filtered)
so_CH_DEV_filtered <- RunUMAP(object = so_CH_DEV_filtered, dims = 1:20)
```

-----------------------------------------------------------------------------------------------------------------------------------------
# pK Identification (no ground truth)
# pK parameter is one of three necessary parameters for doubletfinder
# pK: doubletfinder is very dependant on this, must find the optimal pK value 
# pN: number of artifical doublets ; default : 0.25 (25%) -- doubletfinder performance is not dependant on pN
# expected number of doublets: from user guide of 10x reagents 

```{r}
# paramsSweep introduces artificial doublets in various proportions
# output is a proportion of artificial neighbors for each pK and pN value 
sweep.res.list_so_CH_DEV_filtered <- paramSweep(so_CH_DEV_filtered, PCs = 1:20, sct = FALSE)
# summarizes results of above command
sweep.stats_so_CH_DEV_filtered <- summarizeSweep(sweep.res.list_so_CH_DEV_filtered, GT = FALSE)
# computes a metric so we can find an optimal pk value 
sweep.stats_so_CH_DEV_filtered

# generates plot 
bcmvn_so_CH_DEV_filtered <- find.pK(sweep.stats_so_CH_DEV_filtered)

# plot with labeled axis, plotting pK value on x-axis and BCmetric on y-axis  
# pK value corresponding to the highest BCmetric value is the most optimal pK value
# In this case : 0.1
ggplot(bcmvn_so_CH_DEV_filtered, aes(pK, BCmetric, group = 1)) +
  geom_point() +
  geom_line()

# Storing optimal pK value to pK variable 
# Filtering for max BCmetric --> returned as a list, choose first value of the list (max, optimal)
pK <- bcmvn_so_CH_DEV_filtered %>% # select the pK that corresponds to max bcmvn to optimize doublet detection
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 
pK <- as.numeric(as.character(pK[[1]]))
# check pK variable (should be the one that corresponds to highest on the graph)
```

# Expected number of doublets:
# - from 10x reagents
# - Number of expected loaded cells, number of recovered cells 

Homotypic doublets:
```{r}
# Annotations are based on cell clusters 
# Add for each cell what cluster they belong to 
annotations <- so_CH_DEV_filtered@meta.data$seurat_clusters

# Use annotations to model homotypic doublets 
# Gives a proportion
homotypic.prop <- modelHomotypic(annotations)         

# Based on # cells loaded and # cells recovered, change the number!
nExp_poi <- round(0.077*nrow(so_CH_DEV_filtered@meta.data))  ## Assuming 7.7% doublet formation rate 

# Adjust for homotypic doublets (subtract from one)
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
# Before see 763 --> 692 adjusted doublets                 
```

# Now, we can run doubletfinder 
```{r}
so_CH_DEV_filtered <- doubletFinder(so_CH_DEV_filtered, 
                                     PCs = 1:20, # my dims are 1:20
                                     pN = 0.25, # default set
                                     pK = pK, # provide optimal pK value through the variable
                                     nExp = nExp_poi.adj, # provide expected # of doublets after homtotypic adjustment
                                     reuse.pANN = FALSE, sct = FALSE) # false because this is the first time running doubletfinder on this sample (dont have a pANN to reuse), also did not run sctransform on the data so false also 
```

# Visualize doublets
```{r}
DimPlot(so_CH_DEV_filtered, reduction = 'umap', group.by = "DF.classifications_0.25_0.1_692")
```
# Number of singlets and doublets
```{r}
table(so_CH_DEV_filtered@meta.data$DF.classifications_0.25_0.1_692)
```

--------------------------------------------------------------------------------------------------------------------------
# Pipeline for streamlining doubletfinder processing

# CH_Diapause
```{r}
# PREPROCESSING
so_CH_DIA[["RNA"]]$counts.Peaks <- NULL
so_CH_DIA[["percent.mt"]] <- PercentageFeatureSet(so_CH_DIA, features = c("KEG92-p13", "KEG92-p12", "KEG92-p11", "KEG92-p10", "KEG92-p09", "KEG92-p08", "KEG92-p07", "KEG92-p06", "KEG92-p05", "KEG92-p04", "KEG92-p03", "KEG92-p02", "KEG92-p01"), assay = 'RNA')
VlnPlot(so_CH_DIA, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3)
so_CH_DIA_filtered <- subset(so_CH_DIA, subset = nFeature_RNA > 200 & nFeature_RNA < 1000 & percent.mt < 10)
VlnPlot(so_CH_DIA_filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
so_CH_DIA_filtered <- NormalizeData(object = so_CH_DIA_filtered)
so_CH_DIA_filtered <- FindVariableFeatures(object = so_CH_DIA_filtered)
so_CH_DIA_filtered <- ScaleData(object = so_CH_DIA_filtered)
so_CH_DIA_filtered <- RunPCA(object = so_CH_DIA_filtered)
ElbowPlot(so_CH_DIA_filtered)
so_CH_DIA_filtered <- FindNeighbors(object = so_CH_DIA_filtered, dims = 1:20) # must be adjusted per sample
so_CH_DIA_filtered <- FindClusters(object = so_CH_DIA_filtered)
so_CH_DIA_filtered <- RunUMAP(object = so_CH_DIA_filtered, dims = 1:20)

# PK IDENTIFICATION
sweep.res.list_so_CH_DIA_filtered <- paramSweep(so_CH_DIA_filtered, PCs = 1:20, sct = FALSE)
sweep.stats_so_CH_DIA_filtered <- summarizeSweep(sweep.res.list_so_CH_DIA_filtered, GT = FALSE)
sweep.stats_so_CH_DIA_filtered
bcmvn_so_CH_DIA_filtered <- find.pK(sweep.stats_so_CH_DIA_filtered)
ggplot(bcmvn_so_CH_DIA_filtered, aes(pK, BCmetric, group = 1)) +
  geom_point() +
  geom_line()
pK <- bcmvn_so_CH_DIA_filtered %>% # select the pK that corresponds to max bcmvn to optimize doublet detection
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 
pK <- as.numeric(as.character(pK[[1]]))

# HOMOTYPIC DOUBLETS
annotations <- so_CH_DIA_filtered@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)         
nExp_poi <- round(0.046*nrow(so_CH_DIA_filtered@meta.data))  # must be adjusted per sample
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
 
# RUN DOUBLET FINDER 
so_CH_DIA_filtered <- doubletFinder(so_CH_DIA_filtered, 
                                     PCs = 1:20, # based on dims
                                     pN = 0.25, # default set
                                     pK = pK, # provide optimal pK value through the variable
                                     nExp = nExp_poi.adj, # provide expected # of doublets after homtotypic adjustment
                                     reuse.pANN = FALSE, sct = FALSE) # false because this is the first time running doubletfinder on this sample (dont have a pANN to reuse), also did not run sctransform on the data so false also 

# VISUALIZE DOUBLETS
DimPlot(so_CH_DIA_filtered, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_172")
table(so_CH_DIA_filtered@meta.data$DF.classifications_0.25_0.3_172)

```

# SG_Development
```{r}
# PREPROCESSING
so_SG_DEV[["RNA"]]$counts.Peaks <- NULL
so_SG_DEV[["percent.mt"]] <- PercentageFeatureSet(so_SG_DEV, features = c("KEG92-p13", "KEG92-p12", "KEG92-p11", "KEG92-p10", "KEG92-p09", "KEG92-p08", "KEG92-p07", "KEG92-p06", "KEG92-p05", "KEG92-p04", "KEG92-p03", "KEG92-p02", "KEG92-p01"), assay = 'RNA')
VlnPlot(so_SG_DEV, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3)
so_SG_DEV_filtered <- subset(so_SG_DEV, subset = nFeature_RNA > 200 & nFeature_RNA < 1200 & percent.mt < 10)
VlnPlot(so_SG_DEV_filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
so_SG_DEV_filtered <- NormalizeData(object = so_SG_DEV_filtered)
so_SG_DEV_filtered <- FindVariableFeatures(object = so_SG_DEV_filtered)
so_SG_DEV_filtered <- ScaleData(object = so_SG_DEV_filtered)
so_SG_DEV_filtered <- RunPCA(object = so_SG_DEV_filtered)
ElbowPlot(so_SG_DEV_filtered)
so_SG_DEV_filtered <- FindNeighbors(object = so_SG_DEV_filtered, dims = 1:20) # must be adjusted per sample
so_SG_DEV_filtered <- FindClusters(object = so_SG_DEV_filtered)
so_SG_DEV_filtered <- RunUMAP(object = so_SG_DEV_filtered, dims = 1:20)

# PK IDENTIFICATION
sweep.res.list_so_SG_DEV_filtered <- paramSweep(so_SG_DEV_filtered, PCs = 1:20, sct = FALSE)
sweep.stats_so_SG_DEV_filtered <- summarizeSweep(sweep.res.list_so_SG_DEV_filtered, GT = FALSE)
sweep.stats_so_SG_DEV_filtered
bcmvn_so_SG_DEV_filtered <- find.pK(sweep.stats_so_SG_DEV_filtered)
ggplot(bcmvn_so_SG_DEV_filtered, aes(pK, BCmetric, group = 1)) +
  geom_point() +
  geom_line()
pK <- bcmvn_so_SG_DEV_filtered %>% # select the pK that corresponds to max bcmvn to optimize doublet detection
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 
pK <- as.numeric(as.character(pK[[1]]))

# HOMOTYPIC DOUBLETS
annotations <- so_SG_DEV_filtered@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)         
nExp_poi <- round(0.077*nrow(so_SG_DEV_filtered@meta.data))  # must be adjusted per sample
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
 
# RUN DOUBLET FINDER 
so_SG_DEV_filtered <- doubletFinder(so_SG_DEV_filtered, 
                                     PCs = 1:20, # based on dims
                                     pN = 0.25, # default set
                                     pK = pK, # provide optimal pK value through the variable
                                     nExp = nExp_poi.adj, # provide expected # of doublets after homtotypic adjustment
                                     reuse.pANN = FALSE, sct = FALSE) # false because this is the first time running doubletfinder on this sample (dont have a pANN to reuse), also did not run sctransform on the data so false also 


# VISUALIZE DOUBLETS
DimPlot(so_SG_DEV_filtered, reduction = 'umap', group.by = "DF.classifications_0.25_0.16_1262")
table(so_SG_DEV_filtered@meta.data$DF.classifications_0.25_0.16_1262)
```

```{r}
# PREPROCESSING
so_SG_DIA[["RNA"]]$counts.Peaks <- NULL
so_SG_DIA[["percent.mt"]] <- PercentageFeatureSet(so_SG_DIA, features = c("KEG92-p13", "KEG92-p12", "KEG92-p11", "KEG92-p10", "KEG92-p09", "KEG92-p08", "KEG92-p07", "KEG92-p06", "KEG92-p05", "KEG92-p04", "KEG92-p03", "KEG92-p02", "KEG92-p01"), assay = 'RNA')
VlnPlot(so_SG_DIA, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3)
so_SG_DIA_filtered <- subset(so_SG_DIA, subset = nFeature_RNA > 200 & nFeature_RNA < 1000 & percent.mt < 10)
VlnPlot(so_SG_DIA_filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
so_SG_DIA_filtered <- NormalizeData(object = so_SG_DIA_filtered)
so_SG_DIA_filtered <- FindVariableFeatures(object = so_SG_DIA_filtered)
so_SG_DIA_filtered <- ScaleData(object = so_SG_DIA_filtered)
so_SG_DIA_filtered <- RunPCA(object = so_SG_DIA_filtered)
ElbowPlot(so_SG_DIA_filtered)
so_SG_DIA_filtered <- FindNeighbors(object = so_SG_DIA_filtered, dims = 1:20) # must be adjusted per sample
so_SG_DIA_filtered <- FindClusters(object = so_SG_DIA_filtered)
so_SG_DIA_filtered <- RunUMAP(object = so_SG_DIA_filtered, dims = 1:20)

# PK IDENTIFICATION
sweep.res.list_so_SG_DIA_filtered <- paramSweep(so_SG_DIA_filtered, PCs = 1:20, sct = FALSE)
sweep.stats_so_SG_DIA_filtered <- summarizeSweep(sweep.res.list_so_SG_DIA_filtered, GT = FALSE)
sweep.stats_so_SG_DIA_filtered
bcmvn_so_SG_DIA_filtered <- find.pK(sweep.stats_so_SG_DIA_filtered)
ggplot(bcmvn_so_SG_DIA_filtered, aes(pK, BCmetric, group = 1)) +
  geom_point() +
  geom_line()
pK <- bcmvn_so_SG_DIA_filtered %>% # select the pK that corresponds to max bcmvn to optimize doublet detection
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 
pK <- as.numeric(as.character(pK[[1]]))

# HOMOTYPIC DOUBLETS
annotations <- so_SG_DIA_filtered@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)         
nExp_poi <- round(0.077*nrow(so_SG_DIA_filtered@meta.data))  # must be adjusted per sample
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
 
# RUN DOUBLET FINDER 
so_SG_DIA_filtered <- doubletFinder(so_SG_DIA_filtered, 
                                     PCs = 1:20, # based on dims
                                     pN = 0.25, # default set
                                     pK = pK, # provide optimal pK value through the variable
                                     nExp = nExp_poi.adj, # provide expected # of doublets after homtotypic adjustment
                                     reuse.pANN = FALSE, sct = FALSE) # false because this is the first time running doubletfinder on this sample (dont have a pANN to reuse), also did not run sctransform on the data so false also 


# VISUALIZE DOUBLETS
DimPlot(so_SG_DIA_filtered, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_687")
table(so_SG_DIA_filtered@meta.data$DF.classifications_0.25_0.3_687)
```

