# This script runs a weighted-nearest-neighbor analysis for multiome data


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("AnnotationHub")
```

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ensembldb")
```


```{r}
library(Seurat)
library(Signac)
library(ensembldb)
```


```{r}
gtffile <- "/Users/aakritisingh/Downloads/file.gtf"
DB <- ensDbFromGtf(gtf = gtffile)

## load the DB file directly
EDB <- EnsDb(DB)

## alternatively, build the annotation package
## and finally we can generate the package
makeEnsembldbPackage(ensdb = DB, version = "110", maintainer = "A Singh", author = "A Singh")
```

```{r}
edbbb <- EnsDb(DB)
edbbb
```

Read in h5 (rna) and fragments (atac) files
```{r}
counts <- Read10X_h5("~/multiome/CH_DEV_filtered_feature_bc_matrix.h5")
fragpath <- "~/multiome/CH_DEV_atac_fragments.tsv.gz"
```

get genome annotation for killifish
```{r}
annotation <- GetGRangesFromEnsDb(ensdb = edbbb)
```

```{r}
seqlevels(annotation) <- paste0('chr', seqlevels(annotation))
```

```{r}
multiome <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)
```

Needs to include the index (.tbi) file as well, with the SAME NAME!
```{r}
multiome[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotation
)
```

```{r}
rm(counts)
rm(fragpath)
```


```{r}
multiome
```

```{r}
DefaultAssay(multiome) <- "ATAC"

multiome <- NucleosomeSignal(multiome)
#multiome <- TSSEnrichment(multiome)

VlnPlot(
  object = multiome,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)
```

```{r}
# call peaks using MACS2
peaks <- CallPeaks(multiome, macs2.path = "/Users/aakritisingh/miniforge3/bin/macs3")

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)
```
```{r}
# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(multiome),
  features = peaks,
  cells = colnames(multiome)
)
```

```{r}
# create a new assay using the MACS2 peak set and add it to the Seurat object
multiome[["peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragpath,
  annotation = annotation
)
```

```{r}
DefaultAssay(multiome) <- "RNA"
multiome <- SCTransform(multiome)
multiome <- RunPCA(multiome)
```

```{r}
nrow(x = multiome@scale.data)
```


```{r}
DefaultAssay(multiome) <- "peaks"
multiome <- FindTopFeatures(multiome, min.cutoff = 5)
multiome <- RunTFIDF(multiome)
multiome <- RunSVD(multiome, n = 1)
```
```{r}
ncol(multiome)
nrow(multiome)
```

```{r}
# build a joint neighbor graph using both assays
multiome <- FindMultiModalNeighbors(
  object = multiome,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:15, 2:15),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
multiome <- RunUMAP(
  object = multiome,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)

DimPlot(multiome, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()
```
